File: frontend/lib/models/interaction.dart
// Backend accepts exactly these values for `kind` (see schemas/interaction.py).
const interactionKinds = ['call', 'meeting', 'email', 'note', 'other'];

class Interaction {
  const Interaction({
    required this.id,
    required this.kind,
    required this.subject,
    this.notes,
    required this.occurredAt,
    this.durationMinutes,
    this.done = false,
    this.tags = const [],
    this.contactIds = const [],
    this.organizationIds = const [],
    this.dealIds = const [],
    this.projectIds = const [],
  });

  final String id;
  final String kind;
  final String subject;
  final String? notes;

  /// Past = logged activity, future = planned mail or meeting.
  final DateTime occurredAt;
  final int? durationMinutes;
  final bool done;
  final List<String> tags;

  // What the interaction was about. Contacts used to be the only link, so
  // "every call about this deal" had no answer. A kickoff call is with people,
  // about a deal, under a project — all three at once.
  final List<String> contactIds;
  final List<String> organizationIds;
  final List<String> dealIds;
  final List<String> projectIds;

  /// How many records this interaction is attached to, all kinds together.
  int get linkCount =>
      contactIds.length +
      organizationIds.length +
      dealIds.length +
      projectIds.length;

  bool get isPlanned => occurredAt.isAfter(DateTime.now());

  /// Planned, its time has passed, and it was never confirmed as happened.
  bool get isOverdue => !isPlanned && !done;

  factory Interaction.fromJson(Map<String, dynamic> json) => Interaction(
        id: json['id'] as String,
        kind: json['kind'] as String,
        subject: json['subject'] as String,
        notes: json['notes'] as String?,
        occurredAt: DateTime.parse(json['occurred_at'] as String),
        durationMinutes: json['duration_minutes'] as int?,
        done: json['done'] as bool? ?? false,
        tags: (json['tags'] as List<dynamic>?)?.cast<String>() ?? [],
        contactIds:
            (json['contact_ids'] as List<dynamic>?)?.cast<String>() ?? [],
        organizationIds:
            (json['organization_ids'] as List<dynamic>?)?.cast<String>() ?? [],
        dealIds: (json['deal_ids'] as List<dynamic>?)?.cast<String>() ?? [],
        projectIds:
            (json['project_ids'] as List<dynamic>?)?.cast<String>() ?? [],
      );
}

File: backend/app/schemas/interaction.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

InteractionKind = Literal["call", "meeting", "email", "note", "other"]


class InteractionLinks(BaseModel):
    """What an interaction was about. All independent, all optional.

    Contacts were the only link an interaction could have, so "every call about
    this deal" had no answer. A kickoff call is with people, about a deal, under
    a project — all three at once.
    """

    contact_ids: list[UUID] = []
    organization_ids: list[UUID] = []
    deal_ids: list[UUID] = []
    project_ids: list[UUID] = []


class InteractionCreate(InteractionLinks):
    kind: InteractionKind = "note"
    subject: str
    notes: str | None = None
    # Future timestamps are planned interactions, past ones the activity log.
    occurred_at: datetime
    duration_minutes: int | None = None
    done: bool = False
    tags: list[str] = []


class InteractionUpdate(BaseModel):
    kind: InteractionKind | None = None
    subject: str | None = None
    notes: str | None = None
    occurred_at: datetime | None = None
    duration_minutes: int | None = None
    done: bool | None = None
    tags: list[str] | None = None
    # Whole-list replacement; an empty list detaches everything.
    contact_ids: list[UUID] | None = None
    organization_ids: list[UUID] | None = None
    deal_ids: list[UUID] | None = None
    project_ids: list[UUID] | None = None


class InteractionRead(InteractionLinks):
    id: UUID
    kind: InteractionKind
    subject: str
    notes: str | None
    occurred_at: datetime
    duration_minutes: int | None
    done: bool
    tags: list[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
